verifiers/proofofsql/src/lib.rs

```diff
@@ -0,0 +1,214 @@
+// Copyright 2024, Horizen Labs, Inc.
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use codec::{Decode, Encode, MaxEncodedLen};
+use educe::Educe;
+use errors::ErrorWrapper;
+use frame_support::weights::Weight;
+use hp_verifiers::{Cow, Verifier, VerifyError};
+use proof_of_sql_verifier::VerificationKey;
+use scale_info::TypeInfo;
+use sp_core::Get;
+use sp_std::{marker::PhantomData, vec::Vec};
+
+pub mod benchmarking;
+mod errors;
+mod verifier_should;
+mod weight;
+pub use weight::WeightInfo;
+
+// Here educe is used for Clone, Debug, and PartialEq to work around
+// a long-standing compiler bug https://github.com/rust-lang/rust/issues/26925
+#[derive(Educe, Encode, Decode, TypeInfo)]
+#[educe(Clone, Debug, PartialEq)]
+#[scale_info(skip_type_params(T))]
+pub struct Vk<T>(Vec<u8>, PhantomData<T>);
```

Comment:
Is it necessary to have this dependency from T here (I guess that's the one causing issues with the derivative as well ?). The code related to validate size can be moved inside the trait implementation of the verifier, or maybe you can avoid over-abstracting the MaxNu parameter. Anyway it's not important

Reply:
The dependency from `T` is required for `impl<T: Config> MaxEncodedLen for Vk<T>`: if `Vk` does not have access to `Config::LargestMaxNu`, then it doesn't know how to compute its maximum size.

Reply:
Can't you put this code in the validate_vk function directly ?

Reply:
`MaxEncodedLen` trait is required for saving the `Vk` into storage.